File: Programmers/Python/Level2/Kakao/구현/레벨2. 오픈채팅방.py
def solution(record):
    answer = []

    DB = dict()

    for rec in record:
        rec=list(map(str,rec.split()))
        cmd=rec[0]
        if cmd=="Leave":
            continue
        uid=rec[1]
        nick=rec[2]
        DB[uid]=nick

    in_msg="님이 들어왔습니다."
    out_msg="님이 나갔습니다."

    for rec in record:
        rec=list(map(str,rec.split()))
        cmd=rec[0]
        uid=rec[1]
        if cmd=="Leave":
            answer.append("".join(DB[uid]+out_msg))
        elif cmd=="Enter":
            answer.append("".join(DB[uid]+in_msg))


    return answer

record = ["Enter uid1234 Muzi", "Enter uid4567 Prodo","Leave uid1234","Enter uid1234 Prodo","Change uid4567 Ryan"]

print(solution(record))

File: CodingTest_Java/src/programmers/level02/레벨2.오픈 채팅방.java
package programmers.level02;
import java.util.*;

// 오픈 채팅방
public class OpenChatRoom_carnival77 {
//    class Solution {

        StringTokenizer st = new StringTokenizer("");
        HashMap<String,String> ID_Name = new HashMap<>();

        public String[] solution(String[] record) {
            for(int i=record.length-1;i>=0;i--){
                st = new StringTokenizer(record[i]);
                if(st.nextToken().equals("Leave")) {
                    continue;
                }
                String id = st.nextToken();
                String name = st.nextToken();
                if(!ID_Name.containsKey(id)) {
                    ID_Name.put(id,name);
                }
            }
            ArrayList<String> ans = new ArrayList<>();
            String in = "님이 들어왔습니다.";
            String out = "님이 나갔습니다.";

            for(int i=0;i<record.length;i++){
                st = new StringTokenizer(record[i]);
                String s = st.nextToken();
                if(s.equals("Enter")) {
                    ans.add(ID_Name.get(st.nextToken())+in);
                }
                else if(s.equals("Leave")) {
                    ans.add(ID_Name.get(st.nextToken())+out);
                }
            }
            return ans.toArray(new String[ans.size()]);

        }
//    }

    public static void main(String[] args) {

        String[] record = {"Enter uid1234 Muzi",
                "Enter uid4567 Prodo",
                "Leave uid1234",
                "Enter uid1234 Prodo",
                "Change uid4567 Ryan"};

        OpenChatRoom_carnival77 sol = new OpenChatRoom_carnival77();
        sol.solution(record);
    }
}
